### PR TITLE
STM32WB0: add ADC support

### PR DESCRIPTION
**Pull request by mathieuchopstm**
Add ADC driver for STM32WB0 series hardware. Updates the boards' YAML accordingly and add a test overlay for Nucleo-WB09KE, so that correct behavior can be verified.

The STM32WB0 ADC uses fixed Vinput references that are independent from Vdd.
For compatibility with the Zephyr ADC API model:
* `ADC_REF_INTERNAL` = `ADC_REF_VDD_1` = 3.6V (reflected by `api->ref_internal`)
* `ADC_REF_VDD_1_2` = 2.4V
  * Should be `ADC_REF_VDD_2_3` instead, but this value doesn't exist in API...
* `ADC_REF_VDD_1_3` = 1.2V

Channel numbers correspond to the value used in `SEQ_x` registers (c.f. RM0505 §12.6.6)

### CODE REVIEW

**Review by gautierg-st — APPROVED** [review]

File: dts/arm/st/wb0/stm32wb0.dtsi [hunk 1]
```diff
@@ -217,6 +217,22 @@
...
+			compatible = "st,stm32wb0-adc";
+			reg = <0x41006000 256>;
+			/* On STM32WB0, the two ADC clock domains (ANALOG / DIGITAL)
+			 * can be controlled independently. Expose this feature to
+			 * the driver by having two `clocks` property entries:
+			 *  - first entry is digital part of ADC block (always-on)
+			 *  - second entry is analog part of ADC block (on-demand)
+			 */
+			clocks = <&rcc STM32_CLOCK_BUS_APB1 (1 << 4)>,
+				 <&rcc STM32_CLOCK_BUS_APB1 (1 << 5)>;
```
Comment (gautierg-st) [inline dts/arm/st/wb0/stm32wb0.dtsi:231]: Now that #79683 has been merged, this node should be updated

**Review by gautierg-st — CHANGES_REQUESTED** [review]

File: dts/arm/st/wb0/stm32wb0.dtsi (hunk 1 above)
Comment (mathieuchopstm) [reply dts/arm/st/wb0/stm32wb0.dtsi:231]: Done. Also added a commit to cleanup the missed nodes (I2C/SPI).

**Review by gautierg-st — DISMISSED** [review]

File: dts/arm/st/wb0/stm32wb0.dtsi [hunk 2]
```diff
@@ -217,6 +217,22 @@
...
+			compatible = "st,stm32wb0-adc";
+			reg = <0x41006000 256>;
+			/* On STM32WB0, the two ADC clock domains (ANALOG / DIGITAL)
+			 * can be controlled independently. Expose this feature to
+			 * the driver by having two `clocks` property entries:
+			 *  - first entry is digital part of ADC block (always-on)
+			 *  - second entry is analog part of ADC block (on-demand)
+			 */
+			clocks = <&rcc STM32_CLOCK(APB1, 4)>,
+				 <&rcc STM32_CLOCK(APB1, 5)>;
```
Comment (erwango) [inline dts/arm/st/wb0/stm32wb0.dtsi:231]: Nit (not blocking): clock-names could be handy in this case.

**Review by erwango — DISMISSED** [review]

**Comment by erwango** [comment]
@anangl Do you want to review this yourself ? Or can I take assignmentship ?
Ideally, we'd like to see this in DV4.0

**Comment by anangl** [comment]
> @anangl Do you want to review this yourself ? Or can I take assignmentship ?

Yes, please.

**Comment by aescolar** [comment]
@mathieuchopstm please rebase

**Review by erwango — APPROVED** [review]

**Review by gautierg-st — APPROVED** [review]

**Pull request merged**